common/network/queues/concurrent_queue.c.inc

```diff
@@ -0,0 +1,110 @@
+           CONCURRENT_QUEUE_OF_##TYPE##_BACK,   \
+           CONCURRENT_QUEUE_OF_##TYPE##_EMPTY,  \
+           CONCURRENT_QUEUE_OF_##TYPE##_SIZE,   \
+           CONCURRENT_QUEUE_OF_##TYPE##_PUSH,   \
+           CONCURRENT_QUEUE_OF_##TYPE##_POP};
+
+#define DEFINE_FUNCTION_FRONT_CONCURRENT_QUEUE_OF(TYPE)    \
+  TYPE *CONCURRENT_QUEUE_OF_##TYPE##_FRONT(                \
+      struct concurrent_queue_of_##TYPE##_s *const self) { \
+    TYPE *data = NULL;                                     \
+    if (self == NULL) return NULL;                         \
+    lock_handle_lock(&self->lock);                         \
+    if (self->front) data = &self->front->data;            \
+    lock_handle_unlock(&self->lock);                       \
+    return data;                                           \
+  }
+
+#define DEFINE_FUNCTION_BACK_CONCURRENT_QUEUE_OF(TYPE)     \
+  TYPE *CONCURRENT_QUEUE_OF_##TYPE##_BACK(                 \
+      struct concurrent_queue_of_##TYPE##_s *const self) { \
+    TYPE *data = NULL;                                     \
+    if (self == NULL) return NULL;                         \
+    lock_handle_lock(&self->lock);                         \
+    if (self->back) data = &self->back->data;              \
+    lock_handle_unlock(&self->lock);                       \
+    return data;                                           \
+  }
+
+#define DEFINE_FUNCTION_EMPTY_CONCURRENT_QUEUE_OF(TYPE)    \
+  bool CONCURRENT_QUEUE_OF_##TYPE##_EMPTY(                 \
+      struct concurrent_queue_of_##TYPE##_s *const self) { \
+    size_t size;                                           \
+    if (self == NULL) return true;                         \
+    lock_handle_lock(&self->lock);                         \
+    size = self->size;                                     \
+    lock_handle_unlock(&self->lock);                       \
+    return size == 0;                                      \
+  }
+
+#define DEFINE_FUNCTION_SIZE_CONCURRENT_QUEUE_OF(TYPE)     \
+  size_t CONCURRENT_QUEUE_OF_##TYPE##_SIZE(                \
+      struct concurrent_queue_of_##TYPE##_s *const self) { \
+    size_t size;                                           \
+    if (self == NULL) return 0;                            \
+    lock_handle_lock(&self->lock);                         \
+    size = self->size;                                     \
+    lock_handle_unlock(&self->lock);                       \
+    return size;                                           \
+  }
+
+#define DEFINE_FUNCTION_PUSH_CONCURRENT_QUEUE_OF(TYPE)                      \
+  concurrent_queue_status_t CONCURRENT_QUEUE_OF_##TYPE##_PUSH(              \
+      struct concurrent_queue_of_##TYPE##_s *const self, TYPE const data) { \
+    concurrent_queue_node_of_##TYPE##_t *node;                              \
+    if (self == NULL) return CONCURRENT_QUEUE_NULL_SELF;                    \
+    node = calloc(1, sizeof(*node));                                        \
```

Comment:
to reduce risk of errors, rather do `sizeof(concurrent_queue_node_of_##TYPE##_t)` ? `sizeof(*node)` is a bit hard to read/parse,
`*node` effectively dereferences a nullptr